### PR TITLE
Redo "DB-11270 some refactorings in Write Pipeline"

### PR DESCRIPTION
**Pull request by martinrupp**
## Description
Some refactorings in Write Pipeline

## How to test
No test necessary, functionality and behavior should stay the same and is covered by existing ITs / QA tests.

### CODE REVIEW

**Comment by martinrupp** [comment]
Redo because was reverted because
`HBaseBulkLoadIT failure is reproducible locally. I am reverting the pull request in master`

**Comment by martinrupp** [comment]
compared to first DB-11270, pulled out changes to SITransactor (bit too risky) and InsertDataSetWriter (not really useful). 

Main change is fix in BulkInsertDataSetWriter which caused a fail of `com.splicemachine.derby.impl.load.HBaseBulkLoadIT#testComputeTableSplitKeys` . This was cause because i deleted some code that i thought was unused variables but actually performed operations.

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/BulkInsertDataSetWriter.java [hunk 1]
```diff
@@ -182,29 +181,44 @@ public TxnView getTxn(){
...
                 allCongloms.add(searchCD.getConglomerateNumber());
                 break;
             }
         }
+        RowAndIndexGenerator rowAndIndexGenerator = getRowAndIndexGenerator(tentativeIndexList);
+        DataSet rowAndIndexes = dataSet.flatMap(rowAndIndexGenerator);
+        DataSet keys = rowAndIndexes.mapPartitions(
+                new RowKeyGenerator(bulkImportDirectory, heapConglom,
+                        indexConglom, isUnique));
+        keys.collect();
```
Comment (hatyo) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/BulkInsertDataSetWriter.java:218]: Is this block of code necessary? It seems that `keys` is not used, and there is no obvious desired side-effect. Could you please check?

Comment (martinrupp) [reply hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/BulkInsertDataSetWriter.java:218]: this code is in the old java file as well, see https://github.com/splicemachine/spliceengine/pull/5451/files#diff-b17ac7253e10a871a3e0c8b4c29689c7cfa082c6910cd30c64d6941697392c42L179-L185 (BulkInsertDataSetWriter.java, lines 179-185). Spotbugs was complaining about `List<String> files` not used, so i removed that code, but turned out this made `com.splicemachine.derby.impl.load.HBaseBulkLoadIT#testComputeTableSplitKeys` fail.

If i understand correctly the `RowKeyGenerator` (line 216) is really generating some data, so `keys.collect()` is really doing something, so i basically reverted this to the state it was before, but moved it to a function.

So yes, this is necessary, and `keys` is used as `keys.collect()` is performing work (writes the row keys).

Comment (hatyo) [reply hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/BulkInsertDataSetWriter.java:218]: I see, this is interesting, thanks for explaining the details.

**Review by hatyo — APPROVED** [review]

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3031/

**Review by jyuanca — APPROVED** [review]

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0

**Comment by cloudspliceci** [comment]
TEST SUCCEEDED +1

**Review by carolp-503 — APPROVED** [review]
+1

**Pull request merged**